contracts/src/StakeTable.sol

```diff
@@ -52,493 +102,352 @@ contract StakeTable is AbstractStakeTable, Ownable, InitializedAt {
-    /// @dev TODO this implementation will be revisited later. See
-    /// https://github.com/EspressoSystems/espresso-sequencer/issues/806
-    /// @dev TODO modify this according to the current spec
-    /// @param amount The amount to deposit
-    /// @return (newBalance, effectiveEpoch) the new balance effective at a future epoch
-    function deposit(uint256 amount) external virtual override returns (uint256, uint64) {
-        Node memory node = nodes[msg.sender];
-
-        if (node.account == address(0)) {
-            revert NodeNotRegistered();
-        }
-
-        if (node.account != msg.sender) {
-            revert Unauthenticated();
-        }
-
-        // A node cannot deposit more tokens while it waiting to register.
-        uint64 _currentEpoch = currentEpoch();
-        if (_currentEpoch <= node.registerEpoch) {
-            revert PrematureDeposit();
-        }
+    /// @notice Delegate to a validator
+    /// @param validator The validator to delegate to
+    /// @param amount The amount to delegate
+    function delegate(address validator, uint256 amount) external virtual {
+        ensureValidatorActive(validator);
 
-        // A node cannot deposit more tokens if an exit request is in progress.
-        if (node.exitEpoch != 0) {
-            revert ExitRequestInProgress();
+        uint256 allowance = token.allowance(msg.sender, address(this));
+        if (allowance < amount) {
+            revert InsufficientAllowance(allowance, amount);
         }
 
-        nodes[msg.sender].balance += amount;
-        SafeTransferLib.safeTransferFrom(ERC20(tokenAddress), msg.sender, address(this), amount);
-
-        emit Deposit(msg.sender, uint256(amount));
+        validators[validator].delegatedAmount += amount;
+        delegations[validator][msg.sender] += amount;
 
-        uint64 effectiveEpoch = _currentEpoch + 1;
+        SafeTransferLib.safeTransferFrom(token, msg.sender, address(this), amount);
 
-        return (nodes[msg.sender].balance, effectiveEpoch);
+        emit Delegated(msg.sender, validator, amount);
     }
 
-    /// @notice Request to exit from the stake table, not immediately withdrawable!
-    ///
-    /// @dev TODO modify this according to the current spec
-    function requestExit() external virtual override {
-        Node memory node = nodes[msg.sender];
+    /// @notice Undelegate from a validator
+    /// @param validator The validator to undelegate from
+    /// @param amount The amount to undelegate
+    function undelegate(address validator, uint256 amount) external virtual {
+        ensureValidatorActive(validator);
 
-        if (node.account == address(0)) {
-            revert NodeNotRegistered();
+        if (validators[msg.sender].status == ValidatorStatus.Exited) {
+            revert ValidatorAlreadyExited();
         }
 
-        if (node.account != msg.sender) {
-            revert Unauthenticated();
+        uint256 balance = delegations[validator][msg.sender];
```

Comment:
leaving this comment to remind myself to check whether there's a way for a delegator to know how many tokens they delegated

Reply:
there's no way to do so directly from from the state of the contract, but it's possible via events

Reply:
i think this is a good feature for analytics and also so that a user can determine how much they want to `undelegate` - @veilgets thoughts on adding the ability for a delegator to know the total number of tokens delegated? it's currently not an option

Reply:
If we need it for on-chain voting it's a different story but that's a separate discussion.

Reply:
we're actually having a conversation today about on-chain voting so may need to come back to this and create a github issue if necessary